sql: populate pg_catalog.pg_aggregate table

abhishek20123g opened the pull request:
Fixes https://github.com/cockroachdb/cockroach/issues/44616

This commit added virtual Schema and populated pg_catalog.pg_aggregate table,
along with its respective testcases.

Release justification: low-risk functionality addition.

Release note (sql change): This PR adds virtual Schema and populate
pg_catalog.pg_aggregate table.

abhishek20123g committed:
sql: populate pg_catalog.pg_aggregate table

Fixes https://github.com/cockroachdb/cockroach/issues/44616

This commit added virtual Schema and populated pg_catalog.pg_aggregate table,
along with its respective testcases.

Release justification: low-risk functionality addition.

Release note (sql change): This PR adds virtual Schema and populate
pg_catalog.pg_aggregate table.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/48126)
<!-- Reviewable:end -->

abhishek20123g commented:
Small Overview:
1. As far as I understand aggregated function in crdb is created differently from psql as we don't uses `CREATE AGGREGATE` therefore I think `aggtransfn`, 
  `aggtransfn`, `aggcombinefn` `aggserialfn`, `aggdeserialfn` and .. should be 
populated as `NULL`.
2. I think we could populate `aggtransspace` by somehow using https://github.com/cockroachdb/cockroach/blob/03e5567988e65d3d3f14066c33585efc2ab33cff/pkg/sql/sem/tree/aggregate_funcs.go#L43 but I am not sure should we populate it with it.
3. The logic used for populating `aggsortop` is non-dynamic in nature though I had cross-checked with psql it is compatible with it. As I was not able to figure out how could be determined sort operator for given aggregate function, therefore, I had to hard code it. If there is more proper way to do this I will update as per you review.

yuzefovich commented:
@rohany I think you'd be the best person to review this

jordanlewis reviewed (commented):
Thanks for the submission! I'm fine with leaving `aggtransspace` as null.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48126#-:-M63NoXR1C2b8wm7fRg1:bghprvw)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @abhishek20123g, @jordanlewis, @rohany, and @solongordon)

___
*[pkg/sql/pg_catalog.go, line 2988 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/48126#-M63MYCf7_A89o5un00a:-M63MYCf7_A89o5un00b:b-gxz9hn) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e73427f85ef5b96c0657c1922e40232ffee07f0/pkg/sql/pg_catalog.go#L2988)):*
> ```Go
> 						returnType := tree.NewDOid(tree.DInt(oid.T_bool))
> 						switch name {
> 						case "max", "bit_or":
> ```

bit_or and bit_and don't have sort ops, but bool_and and bool_or do.
___
*[pkg/sql/pg_catalog.go, line 2996 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/48126#-M63JmzD3mGW8r3P6HN3:-M63JmzD3mGW8r3P6HN4:b-qujrfi) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e73427f85ef5b96c0657c1922e40232ffee07f0/pkg/sql/pg_catalog.go#L2996)):*
> ```Go
> 					err := addRow(
> 						h.BuiltinOid(name, &overload).AsRegProc(name), // aggfnoid
> 						tree.NewDString("n"),                          // aggkind
> ```

I'm not sure exactly why, but I think this needs to vary - `rank`, `percent_rank`, `cume_dist`, and `dense_rank` must be `h`, and all of the `percentile` ones need to be `o`.
___
*[pkg/sql/pg_catalog.go, line 2997 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/48126#-M63NfLD4C2JlWwvj7Mi:-M63NfLD4C2JlWwvj7Mj:b3m3z3y) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e73427f85ef5b96c0657c1922e40232ffee07f0/pkg/sql/pg_catalog.go#L2997)):*
> ```Go
> 						h.BuiltinOid(name, &overload).AsRegProc(name), // aggfnoid
> 						tree.NewDString("n"),                          // aggkind
> 						tree.NewDInt(0),                               // aggnumdirectargs
> ```

Use `tree.DZero` 
___
*[pkg/sql/pg_catalog.go, line 2998 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/48126#-M63N7m95stoY1bV-ujF:-M63N7m95stoY1bV-ujG:bnnli89) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e73427f85ef5b96c0657c1922e40232ffee07f0/pkg/sql/pg_catalog.go#L2998)):*
> ```Go
> 						tree.NewDString("n"),                          // aggkind
> 						tree.NewDInt(0),                               // aggnumdirectargs
> 						tree.DNull,                                    // aggtransfn
> ```

That ones that say (zero if none) in the docs should be DZero, not null, I think. Refer to this: https://www.postgresql.org/docs/12/catalog-pg-aggregate.html
___
*[pkg/sql/pg_catalog.go, line 3006 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/48126#-M63NG418ioTBtxSdsky:-M63NG418ioTBtxSdskz:b-ay1om9) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e73427f85ef5b96c0657c1922e40232ffee07f0/pkg/sql/pg_catalog.go#L3006)):*
> ```Go
> 						tree.DNull,                                    // aggminvtransfn
> 						tree.DNull,                                    // aggmfinalfn
> 						tree.DNull,                                    // aggfinalextra
> ```

This and the next should be DFalse, not null.


<!-- Sent from Reviewable.io -->

abhishek20123g commented:
Thanks for the review @jordanlewis 

Made Changes as you suggested :blush: 
1. Updated from `tree.DNull` to `tree.DBoolFalse`.
2. >  bit_or and bit_and don't have sort ops, but bool_and and bool_or do.

   My bad I had read it wrong, Corrected it.

3. Updated `aggkind` and `aggnumdirectargs` as per the function name `rank`, `percentile_` ... 
4. For `aggtransfn` and other made them look as they were in psql, when no functions is present.
```
regprocForZeroOid := oidZero.AsRegProc("-")
```
```
abhishek=# select 0::regproc;
 regproc 
---------
 -
(1 row)
```
5. Not sure how these changes breaking testcase in `testdata/logic_test/orms`though I had updated them.


6. There are some inconsistency between 
    https://www.postgresql.org/docs/12/catalog-pg-aggregate.html and 
    https://www.postgresql.org/docs/9.6/catalog-pg-aggregate.html 
    Though I had followed 9.6 as other schemas are also following it too.

jordanlewis commented on pkg/sql/logictest/testdata/logic_test/pg_catalog:
```diff
@@ -2022,6 +2025,31 @@ SELECT objoid, classoid, provider, label FROM pg_catalog.pg_shseclabel
+oid  oprname  aggsortop
+
+# Check whether correct operator's oid is set for min, bool_and and every.
+query OTO colnames
```
These tests look good but I don't quite understand why they don't return any values. A better test would probably produce the oid for min, bool_and and every. same goes for the above.

abhishek20123g replied:
Sure

jordanlewis reviewed (commented):
Looking good!

abhishek20123g commented on pkg/sql/sqlbase/constants.go:
```diff
@@ -142,6 +142,7 @@ const (
 	PgCatalogPreparedStatementsTableID
 	PgCatalogPreparedXactsTableID
 	PgCatalogProcTableID
+	PgCatalogAggregateTableID
```
Sure @jordanlewis 
along with that, I have updated `PgCatalogID` as it was also not in order.

jordanlewis replied:
Sorry, but that one should stay where it was, at the top of the PGCatalog definitions, to follow the convention in the file.

abhishek20123g replied:
Sorry, my bad I should have asked :blush: 
Corrected

jordanlewis reviewed (approved):
Thank you @abhishek20123g!

bors r+

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48126)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @abhishek20123g, @jordanlewis, @rohany, and @solongordon)

___
*[pkg/sql/pg_catalog.go, line 2997 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/48126#-M6StnLG71S5R26DnGDN:-M6StnLH6dWcNXRsQqFc:bi9uw12) ([raw file](https://github.com/cockroachdb/cockroach/blob/4047d45256b8cc7291d0167916115066d4f09f90/pkg/sql/pg_catalog.go#L2997)):*
> ```Go
> 
> 							// Cases to determine aggregate kind.
> 							case "rank", "percent_rank", "cume_dict", "dense_rank":
> ```

`s/cume_dict/cume_dist/g`.


<!-- Sent from Reviewable.io -->

abhishek20123g commented:
Sorry, @yuzefovich,
I am unable to understand the issue you are pointing at can you give me briefer at it.
Thanks

yuzefovich commented:
I meant that there is a typo in the code, `cume_dict` - it should be `cume_dist`, and the syntax "s/foo/bar/g" means replace "foo" with "bar".

abhishek20123g commented:
> I meant that there is a typo in the code, cume_dict - it should be cume_dist, and the syntax "s/foo/bar/g" means replace "foo" with "bar".

@yuzefovich thank you for identifying the typo.
I have created a separate PR https://github.com/cockroachdb/cockroach/pull/48351 as you suggested to correct the typo.
